File: unmastered/ctci_problems/trees_and_graphs/4.6.js
/*
	given node.  
	If right exists:
		set right to ret
		while left of ret is true set ret to left
	else
		set currNode to node
		set parent to currNode.parent
		while parent exists
			if parent.left is child then return parent
			set currNode to parent 
			set parent to currNode.parent

		return parent

*/

class Node{
	constuctor(val){
		this.val = val;
		this.left = undefined;
		this.right = undefined;
		this.parent = undefined;
	}
}

Node.prototype.addLeft(child){
	this.left = child;
	child.parent = this;
}


Node.prototype.addRight(child){
	this.right = child;
	child.parent = this;
}

function successor(node){
	if (node.right){
		let ret = node.right;
		while(ret.left){
			ret = ret.left;
		}
		return ret;
	} else {
		let currNode = node;
		let parent = currNode.parent;
		while(parent){
			if (parent.left === currNode){
				return parent;
			}
			currNode = parent;
			parent = currNode.parent;
		}
		return parent;
	}
}

File: unmastered/ctci_problems/trees_and_graphs/4.6.py
class Node(object):
	def __init__(self, val):
		self.parent = None
		self.left = None
		self.right = None
		self.val = val

def successor(node):
	if not node.right is None:
		return getLowestChild(node.right);
	else:
		return getFirstParentOfLeftChild(node);

def getFirstParentOfLeftChild(node):
	while (not node.parent is None):
		if node.parent.left == node:
			return node.parent
		else:
			node = node.parent
	return None

def getLowestChild(node):
	currNode = node;
	while (not currNode.left is None):
		currNode = currNode.left
	return currNode;
